apps/server/src/shared/domain/rules/submission.rule.spec.ts

```diff
@@ -0,0 +1,300 @@
+
+							const result = submissionRule.hasPermission(student, submission, {
+								action: Actions.write,
+								requiredPermissions: [permission],
+							});
+
+							expect(result).toBe(false);
+						});
+					});
+				});
+
+				describe('when action is "read"', () => {
+					describe('when user is in same course and submissions are public', () => {
+						const setup = () => {
+							const permission = 'a' as Permission;
+							const student = buildStudent(permission);
+							const student2 = buildStudent(permission);
+							const course = courseFactory.build({ students: [student, student2] });
+							const task = taskFactory.build({ course, publicSubmissions: true });
+							const submission = submissionFactory.build({ task, student: student2 });
+							return { student, submission, permission };
+						};
+
+						it('should return true', () => {
+							const { student, submission, permission } = setup();
+
+							const result = submissionRule.hasPermission(student, submission, {
+								action: Actions.read,
+								requiredPermissions: [permission],
+							});
+
+							expect(result).toBe(true);
+						});
+					});
+
+					describe('when user does not have write access nor are submissions public', () => {
+						const setup = () => {
+							const permission = 'a' as Permission;
+							const student = buildStudent(permission);
+							const student2 = buildStudent(permission);
+							const course = courseFactory.build({ students: [student, student2] });
+							const task = taskFactory.build({ course });
+							const submission = submissionFactory.build({ task, student: student2 });
+							return { student, submission, permission };
+						};
+
+						it('should return false', () => {
+							const { student, submission, permission } = setup();
+
+							const result = submissionRule.hasPermission(student, submission, {
+								action: Actions.read,
+								requiredPermissions: [permission],
+							});
+
+							expect(result).toBe(false);
+						});
+					});
+				});
+			});
+
+			describe('when user is teacher', () => {
+				describe('when action is "write" (or "read" respectively)', () => {
+					describe('when user is course teacher', () => {
+						const setup = () => {
+							const permission = 'a' as Permission;
+							const teacher = buildTeacher(permission);
+							const course = courseFactory.build({ teachers: [teacher] });
+							const task = taskFactory.build({ course });
+							const submission = submissionFactory.build({ task });
+							return { teacher, submission, permission };
+						};
+
+						it('should return true', () => {
+							const { teacher, submission, permission } = setup();
+
+							const result = submissionRule.hasPermission(teacher, submission, {
+								action: Actions.write,
+								requiredPermissions: [permission],
+							});
+
+							expect(result).toBe(true);
+						});
+					});
+
+					describe('when user is not course teacher', () => {
```

Comment:
Dieser Teacher hat keine Permission oder? Deswegen würde ich das Szenario im when "user does not have required permissions"-Block erwarten.

Reply:
Mit "required permissions" sind hier nur die string-based Permissions gemeint. Den konkreten Zugriff auf das jeweilige Objekt hab ich in Abgrenzung dazu "access" genannt (siehe SubmissionRule). Ich hab versucht, diese Trennung zwischen (allgemeinen) Permissions und (spezifischem) Access durchzuziehen. In den anderen Rules ist das eher durcheinander. Ist aber auch eher als Vorschlag von mir zu sehen.

Reply:
Genau genommen müsste es dann heißen "when user's roles do not have required permissions". Denn ein User kann mehrere Roles haben. Ich bin mir gerade unsicher, ob es nötig ist, dass hier so explizit zu machen.

Reply:
Ich wüsste nicht was dagegen spricht, es explizit ausdrücken. Und für jemand der hier das erste mal drauf schaut ist es wichtig zu verstehen, dass sich die required permissions auf die Rollen beziehen.

Reply:
Stattgegeben :)